File: lib/activity/phone.ts
/**
 * Suivi d'activité sur Android — le pont côté front.
 *
 * Deux mondes, deux façons de mesurer, et il faut le dire clairement parce que
 * tout le reste en découle :
 *
 *  • Sur un POSTE, l'app échantillonne. Elle demande toutes les quelques
 *    secondes quelle fenêtre est devant et construit la journée au fil de l'eau
 *    (cf. lib/activity/engine).
 *  • Sur ANDROID, elle RECONSTRUIT. Le système gèle le WebView dès que tao
 *    passe en arrière-plan : une boucle s'arrêterait exactement quand il y
 *    aurait quelque chose à voir. Mais Android tient déjà le journal des
 *    passages au premier plan, et on le relit à l'ouverture. Rien ne tourne en
 *    fond, rien ne consomme, et il n'y a aucun trou — le temps passé pendant
 *    que tao était fermé est là aussi.
 *
 * Ce que la plateforme ne donne pas : le TITRE de la fenêtre. On sait « Chrome
 * pendant 2 h 10 », jamais quel site. L'interface doit donc parler
 * d'applications et ne rien promettre de plus.
 *
 * Les mesures restent locales à l'appareil, comme sur un poste : le téléphone a
 * son propre magasin de données, donc son propre journal. C'est voulu — mélanger
 * le temps d'écran du téléphone et les heures de travail du Mac dans une même
 * journée donnerait des totaux qui ne veulent rien dire.
 */

import { isTauri } from "@/lib/notify";
import { classifyPhoneApp, type ClassifyRule } from "@/lib/activity/categories";
import type { Segment } from "@/lib/activity/engine";

/** Un passage au premier plan, tel que le système l'a enregistré. */
export interface PhoneSegment {
  packageName: string;
  app: string;
  s: number;
  e: number;
}

export interface UsageAccess {
  /** Vrai si « accès aux données d'utilisation » est accordé. */
  granted: boolean;
  /** Faux hors Android : il n'y a alors rien à autoriser ni à lire. */
  supported: boolean;
}

async function call<T>(command: string, args?: Record<string, unknown>): Promise<T | null> {
  if (!isTauri()) return null;
  try {
    const { invoke } = await import("@tauri-apps/api/core");
    return await invoke<T>(command, args);
  } catch {
    /* Version de l'app antérieure à la commande, ou plateforme sans téléphone :
       l'appelant traite ça comme « non pris en charge ». */
    return null;
  }
}

/** L'appareil sait-il lire son propre usage, et l'autorisation est-elle donnée ? */
export async function usageAccess(): Promise<UsageAccess> {
  const res = await call<UsageAccess>("phone_usage_access");
  return { granted: !!res?.granted, supported: !!res?.supported };
}

/**
 * Ouvre l'écran système d'autorisation.
 *
 * « Accès aux données d'utilisation » est une permission SPÉCIALE : aucune boîte
 * de dialogue ne peut la demander, l'utilisateur la donne écran par écran. On y
 * mène donc, au lieu de faire semblant de la demander.
 */
export async function openUsageSettings(): Promise<void> {
  await call<void>("phone_open_usage_settings");
}

/** Bornes locales d'une journée « YYYY-MM-DD ». */
function dayBounds(date: string): { from: number; to: number } {
  const [y, m, d] = date.split("-").map(Number);
  const start = new Date(y, (m || 1) - 1, d || 1, 0, 0, 0, 0);
  const end = new Date(start);
  end.setDate(end.getDate() + 1);
  // Une journée en cours s'arrête MAINTENANT : demander l'avenir au système ne
  // rend rien, mais fermerait le dernier segment sur une heure qui n'est pas
  // encore arrivée.
  return { from: start.getTime(), to: Math.min(end.getTime(), Date.now()) };
}

/**
 * La journée du téléphone, telle que le système l'a vécue.
 *
 * Rend `null` quand il n'y a rien à lire (hors Android, ou autorisation
 * refusée) : l'appelant garde alors ce qu'il avait, plutôt que d'écraser un
 * journal par une journée vide.
 */
export async function phoneDay(date: string, rules: ClassifyRule[] = []): Promise<Segment[] | null> {
  const { from, to } = dayBounds(date);
  if (to <= from) return null;

  const res = await call<{ segments: PhoneSegment[]; granted: boolean; supported: boolean }>(
    "phone_segments", { from, to }
  );
  if (!res?.supported || !res.granted) return null;

  const out: Segment[] = [];
  for (const raw of res.segments || []) {
    const s = Math.max(from, raw.s);
    const e = Math.min(to, raw.e);
    if (!(e > s)) continue;
    const shown = raw.app || raw.packageName;
    const { category } = classifyPhoneApp(shown, raw.packageName, rules);
    out.push({
      s, e,
      // `app` porte le nom BRUT donné par l'OS — ici le paquet, comme ailleurs
      // le nom de processus. C'est lui qu'une règle « dans l'application » vise.
      app: raw.packageName,
      label: shown,
      title: "",
      cat: category,
    });
  }
  return out.sort((a, b) => a.s - b.s);
}

File: src-tauri/src/phone.rs
/*!
Pont vers le suivi d'activité du téléphone.

Le côté Kotlin est dans `gen/android/app/src/main/java/app/taotrade/desktop/
PhonePlugin.kt`, et son en-tête explique le point essentiel : sur Android on ne
mesure pas, on RECONSTRUIT. `UsageStatsManager` tient déjà le journal des
passages au premier plan ; l'app le relit et en déduit la journée. Aucune boucle
en fond, aucun trou pendant que l'app est fermée.

Ce module ne fait que porter quatre appels d'un monde à l'autre. Il existe sur
TOUTES les plateformes, et c'est voulu : le front interroge la même API partout
et reçoit `supported: false` là où il n'y a pas de téléphone à mesurer, plutôt
que de devoir savoir sur quoi il tourne avant d'appeler.
*/

use serde::{Deserialize, Serialize};
use tauri::{
  plugin::{Builder, TauriPlugin},
  Runtime,
};

/* ─── Ce qui traverse ────────────────────────────────────────────────────── */

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct UsageAccess {
  /// Vrai si « accès aux données d'utilisation » est accordé.
  pub granted: bool,
  /// Faux hors Android : il n'y a alors rien à autoriser.
  #[serde(default)]
  pub supported: bool,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct PhoneSnapshot {
  /// Nom lisible de l'app au premier plan, vide si inconnu.
  pub app: String,
  pub package_name: String,
  /// Écran allumé — le seul « présent » qu'un téléphone connaisse.
  pub screen_on: bool,
  pub granted: bool,
  #[serde(default)]
  pub supported: bool,
}

/// Un passage au premier plan, tel que le système l'a enregistré.
#[derive(Serialize, Deserialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
pub struct PhoneSegment {
  pub package_name: String,
  pub app: String,
  /// Début (ms epoch).
  pub s: i64,
  /// Fin (ms epoch).
  pub e: i64,
}

#[derive(Serialize, Deserialize, Clone, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub struct PhoneSegments {
  pub segments: Vec<PhoneSegment>,
  pub granted: bool,
  #[serde(default)]
  pub supported: bool,
}

/* La fenêtre demandée à Kotlin. Construite uniquement sur Android — ailleurs,
   `phone_segments` répond sans traverser. */
#[cfg_attr(not(target_os = "android"), allow(dead_code))]
#[derive(Serialize, Clone, Debug)]
#[serde(rename_all = "camelCase")]
struct Range {
  from: i64,
  to: i64,
}

/* ─── Android ────────────────────────────────────────────────────────────── */

#[cfg(target_os = "android")]
mod imp {
  use super::*;
  use tauri::plugin::PluginHandle;

  /// Poignée vers la classe Kotlin, posée dans l'état de l'app au démarrage.
  pub struct Phone<R: Runtime>(pub PluginHandle<R>);

  pub fn access<R: Runtime>(p: &Phone<R>) -> Result<UsageAccess, String> {
    p.0
      .run_mobile_plugin::<UsageAccess>("usageAccess", ())
      .map(|mut a| { a.supported = true; a })
      .map_err(|e| e.to_string())
  }

  pub fn open_settings<R: Runtime>(p: &Phone<R>) -> Result<(), String> {
    p.0
      .run_mobile_plugin::<()>("openUsageSettings", ())
      .map_err(|e| e.to_string())
  }

  pub fn snapshot<R: Runtime>(p: &Phone<R>) -> Result<PhoneSnapshot, String> {
    p.0
      .run_mobile_plugin::<PhoneSnapshot>("snapshot", ())
      .map(|mut s| { s.supported = true; s })
      .map_err(|e| e.to_string())
  }

  pub fn segments<R: Runtime>(p: &Phone<R>, from: i64, to: i64) -> Result<PhoneSegments, String> {
    p.0
      .run_mobile_plugin::<PhoneSegments>("segments", Range { from, to })
      .map(|mut s| { s.supported = true; s })
      .map_err(|e| e.to_string())
  }
}

/* ─── Commandes ──────────────────────────────────────────────────────────── */

/* Hors Android, les quatre commandes répondent « non pris en charge » au lieu de
   ne pas exister : une commande absente lève côté front, et le front devrait
   alors deviner la plateforme avant d'appeler. */

#[tauri::command]
pub async fn phone_usage_access<R: Runtime>(app: tauri::AppHandle<R>) -> Result<UsageAccess, String> {
  #[cfg(target_os = "android")]
  {
    use tauri::Manager;
    return imp::access(&*app.state::<imp::Phone<R>>());
  }
  #[cfg(not(target_os = "android"))]
  {
    let _ = app;
    Ok(UsageAccess::default())
  }
}

#[tauri::command]
pub async fn phone_open_usage_settings<R: Runtime>(app: tauri::AppHandle<R>) -> Result<(), String> {
  #[cfg(target_os = "android")]
  {
    use tauri::Manager;
    return imp::open_settings(&*app.state::<imp::Phone<R>>());
  }
  #[cfg(not(target_os = "android"))]
  {
    let _ = app;
    Ok(())
  }
}

#[tauri::command]
pub async fn phone_snapshot<R: Runtime>(app: tauri::AppHandle<R>) -> Result<PhoneSnapshot, String> {
  #[cfg(target_os = "android")]
  {
    use tauri::Manager;
    return imp::snapshot(&*app.state::<imp::Phone<R>>());
  }
  #[cfg(not(target_os = "android"))]
  {
    let _ = app;
    Ok(PhoneSnapshot::default())
  }
}

#[tauri::command]
pub async fn phone_segments<R: Runtime>(
  app: tauri::AppHandle<R>,
  from: i64,
  to: i64,
) -> Result<PhoneSegments, String> {
  #[cfg(target_os = "android")]
  {
    use tauri::Manager;
    return imp::segments(&*app.state::<imp::Phone<R>>(), from, to);
  }
  #[cfg(not(target_os = "android"))]
  {
    let _ = (app, from, to);
    Ok(PhoneSegments::default())
  }
}

/// Branche la classe Kotlin. Sur les autres plateformes, le greffon ne fait rien.
pub fn init<R: Runtime>() -> TauriPlugin<R> {
  Builder::new("phone")
    .setup(|_app, _api| {
      #[cfg(target_os = "android")]
      {
        use tauri::Manager;
        let handle = _api.register_android_plugin("app.taotrade.desktop", "PhonePlugin")?;
        _app.manage(imp::Phone(handle));
      }
      Ok(())
    })
    .build()
}
